### PR TITLE
feat: init support for e2e libs with Playwright

### LINKED ISSUE #116: Consider Playwright support (or playwright-on-rails)
I was able to successfully reuse your library to work with Playwright. The only thing required was [porting Cypress commands to Playwright](https://twitter.com/nerdblogpl/status/1574764868581965825). You should consider either forking this library and changing the naming or creating a common core and make both libraries swappable :)

### PR DESCRIPTION
**Pull request by KhaledEmaraDev**
* Abstract Middleware logic from dependency on Cypress
* Add support for Playwright

Fixes #116

### CODE REVIEW

File: lib/cypress_on_rails/configuration.rb [hunk 1]
```diff
@@ -2,8 +2,8 @@
 
 module CypressOnRails
   class Configuration
-    attr_accessor :cypress_folder
```
Comment (grantspeelman) [inline lib/cypress_on_rails/configuration.rb:5]: this is a breaking change, suggest the method is added again that prints out a deprecation warning and uses install_folder

File: lib/cypress_on_rails/middleware.rb [hunk 2]
```diff
@@ -16,7 +16,7 @@ def initialize(app, command_executor = CommandExecutor, file = ::File)
 
     def call(env)
       request = Rack::Request.new(env)
-      if request.path.start_with?("#{configuration.api_prefix}/__cypress__/command")
+      if request.path.start_with?("#{configuration.api_prefix}/__e2e__/command")
```
Comment (grantspeelman) [inline lib/cypress_on_rails/middleware.rb:19]: another breaking change, we should also check `__cypress__` path and print a deprecation warning.

File: lib/cypress_on_rails/version.rb [hunk 3]
```diff
@@ -1,3 +1,3 @@
 module CypressOnRails
-  VERSION = '1.13.1'.freeze
+  VERSION = '1.14.1'.freeze
```
Comment (grantspeelman) [inline lib/cypress_on_rails/version.rb:2]: I like to keep versioning in a separate commit.

```suggestion
  VERSION = '1.13.1'.freeze
```

File: lib/generators/cypress_on_rails/install_generator.rb [hunk 4]
```diff
@@ -1,45 +1,71 @@
...
   class InstallGenerator < Rails::Generators::Base
-    class_option :cypress_folder, type: :string, default: 'cypress'
     class_option :api_prefix, type: :string, default: ''
+    class_option :install_folder, type: :string, default: 'e2e'
     class_option :install_cypress, type: :boolean, default: true
-    class_option :install_cypress_with, type: :string, default: 'yarn'
+    class_option :install_playwright, type: :boolean, default: false
+    class_option :install_with, type: :string, default: 'yarn'
+    class_option :cypress_folder, type: :string, default: 'cypress'
+    class_option :playwright_folder, type: :string, default: 'playwright'
```
Comment (grantspeelman) [inline lib/generators/cypress_on_rails/install_generator.rb:9]: we should update the `README.md` around this.

**Review by grantspeelman — CHANGES_REQUESTED** [review]
Thx, looks very useful and Great Work so far 👍🏽 

Left a few suggestions
can you also please update the `README.md` and `CHANGELOG.md`

File: lib/cypress_on_rails/command_executor.rb [hunk 5]
```diff
@@ -14,7 +14,7 @@ def self.perform(file,command_options = nil)
     end
 
     def self.load_cypress_helper
-      cypress_helper_file = "#{configuration.cypress_folder}/cypress_helper"
+      cypress_helper_file = "#{configuration.install_folder}/cypress_helper"
```
[the remark this answers is not in the document]
Comment (grantspeelman) [reply lib/cypress_on_rails/command_executor.rb:17]: we can rename to `e2e_helper` but we just need to keep backward compatibility so it should also look for `cypress_helper` and print a deprecation warning about it.

File: lib/generators/cypress_on_rails/templates/spec/playwright.config.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+module.exports = defineConfig({
+  testDir: './playwright/e2e',
+  /* Run tests in files in parallel */
+  fullyParallel: true,
+  /* Fail the build on CI if you accidentally left test.only in the source code. */
+  forbidOnly: !!process.env.CI,
+  /* Retry on CI only */
+  retries: process.env.CI ? 2 : 0,
+  /* Opt out of parallel tests on CI. */
+  workers: process.env.CI ? 1 : undefined,
```
[the remark this answers is not in the document]
Comment (KhaledEmaraDev) [reply lib/generators/cypress_on_rails/templates/spec/playwright.config.js:22]: I believe this is a sensible default?

**Comment by KhaledEmaraDev** [comment]
@grantspeelman @alexeyr-ci I think the PR is ready.

It passes the CI for two envs but not the third for a weird reason.
Will investigate further tomorrow.

File: lib/cypress_on_rails/configuration.rb [hunk 7]
```diff
@@ -2,8 +2,9 @@
...
     attr_accessor :api_prefix
+    attr_accessor :install_folder
+    attr_accessor :cypress_folder
```
Comment (grantspeelman) [inline lib/cypress_on_rails/configuration.rb:7]: I think we rather create a fallback wrapper instead.

```suggestion
  def cypress_folder
    warn "cypress_folder is deprecated, please use install_folder"
    install_folder
  end

  def cypress_folder=(v)
    warn "cypress_folder= is deprecated, please use install_folder"
    self.install_folder = v
  end
```

**Review by grantspeelman — COMMENTED** [review]
Great Work, looking real good 👍🏽 

One last suggestion

File: lib/cypress_on_rails/configuration.rb (hunk 7 above)
Comment (justin808) [reply lib/cypress_on_rails/configuration.rb:7]: Ping @khaledEmaraDev

File: lib/cypress_on_rails/command_executor.rb [hunk 8]
```diff
@@ -13,12 +13,12 @@ def self.perform(file,command_options = nil)
...
-      cypress_helper_file = "#{configuration.cypress_folder}/cypress_helper"
-      if File.exist?("#{cypress_helper_file}.rb")
-        Kernel.require cypress_helper_file
+    def self.load_e2e_helper
+      e2e_helper_file = "#{configuration.install_folder}/e2e_helper.rb"
+      if File.exist?(e2e_helper_file)
+        Kernel.require e2e_helper_file
       else
-        logger.warn "could not find #{cypress_helper_file}.rb"
+        logger.warn "could not find #{e2e_helper_file}"
```
Comment (grantspeelman) [inline lib/cypress_on_rails/command_executor.rb:21]: what happened to the fallback to look for `cypress_helper.rb` so we don't break any existing setups?

Comment (KhaledEmaraDev) [reply lib/cypress_on_rails/command_executor.rb]: Thank you @grantspeelman . Reverted back.

**Review by grantspeelman — APPROVED** [review]
Well Done and Thank you 👍🏽

**Comment by tvongaza** [comment]
Amazing, nice work @KhaledEmaraDev !

File: lib/generators/cypress_on_rails/templates/spec/playwright/support/on-rails.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  Object.keys(options).forEach(key => options[key] === undefined ? delete options[key] : {});
+  const response = await context.post("/__e2e__/vcr/insert", {data: [cassette_name,options]});
+  expect(response.ok()).toBeTruthy();
+  return response.body;
+}
+
+const appVcrEjectCassette = async () => {
+  const context = await contextPromise;
+
+  const response = await context.post("/__cypress__/vcr/eject");
```
Comment (tvongaza) [inline lib/generators/cypress_on_rails/templates/spec/playwright/support/on-rails.js:32]: Should this be

`  const response = await context.post("/__e2e__/vcr/eject");`

Comment (KhaledEmaraDev) [reply lib/generators/cypress_on_rails/templates/spec/playwright/support/on-rails.js:32]: You are right. Don't know how I missed it. Will create a PR for it.
